File: backend/arrlink/api/links.py
"""Links API: browse created hardlinks, delete one, repair missing ones."""
from __future__ import annotations

import os

from fastapi import APIRouter, Depends, HTTPException, Query, Request

from ..core.fsutil import create_link, ensure_dir, inode_of, resolve_fs_fallback
from ..deps import get_db
from ..state import State
from .auth import CurrentUser

router = APIRouter(prefix="/api/links", tags=["links"])


@router.get("")
def list_links(
    _user: CurrentUser,
    db: State = Depends(get_db),
    app_id: int | None = None,
    rule_id: int | None = None,
    status: str | None = Query(default="active"),
    limit: int = Query(default=500, ge=1, le=5000),
    offset: int = Query(default=0, ge=0),
) -> dict:
    """One page of links plus the unpaged `total` for the same filters, so a
    large library can page instead of silently truncating at `limit`."""
    where = " WHERE 1=1"
    params: list = []
    if app_id is not None:
        where += " AND l.app_id=?"
        params.append(app_id)
    if rule_id is not None:
        where += " AND l.rule_id=?"
        params.append(rule_id)
    if status:
        where += " AND l.status=?"
        params.append(status)

    total_row = db.query_one(
        "SELECT COUNT(*) AS c FROM links l" + where, tuple(params)
    )
    total = total_row["c"] if total_row else 0

    rows = db.query(
        "SELECT l.*, r.name AS rule_name, a.name AS app_name, a.type AS app_type "
        "FROM links l LEFT JOIN rules r ON r.id=l.rule_id "
        "LEFT JOIN apps a ON a.id=l.app_id" + where
        + " ORDER BY l.id DESC LIMIT ? OFFSET ?",
        tuple(params) + (limit, offset),
    )
    return {
        "items": [dict(r) for r in rows],
        "total": total,
        "limit": limit,
        "offset": offset,
    }


@router.delete("/{link_id}", status_code=204)
def delete_link(link_id: int, _user: CurrentUser, db: State = Depends(get_db)) -> None:
    from ..core.fsutil import remove_link

    row = db.query_one("SELECT * FROM links WHERE id=?", (link_id,))
    if not row:
        raise HTTPException(404, "link not found")
    r = remove_link(row["dst_path"])
    if not r.ok and r.error:
        raise HTTPException(500, r.error)
    db.execute("UPDATE links SET status='missing' WHERE id=?", (link_id,))
    db.commit()
    db.log_event("info", f"link removed: {row['dst_path']}")


@router.post("/repair")
def repair(
    request: Request,
    _user: CurrentUser,
    db: State = Depends(get_db),
) -> dict:
    """Re-create active links whose dst disappeared but whose source survives.

    Runs in FastAPI's threadpool (sync route), so the O(all-links) stat scan
    never blocks the event loop; the row updates are batched into one commit.
    """
    fixed = 0
    failed = 0
    fallback = resolve_fs_fallback(db, request.app.state.settings.fs_fallback)
    # 'missing' too: repair re-creates links that were removed (by the UI or
    # by the poller) whose source file has since (re)appeared.
    rows = db.query(
        "SELECT id, dst_path, src_path FROM links "
        "WHERE status IN ('active','stale','missing')"
    )
    with db.transaction():
        for row in rows:
            dst = row["dst_path"]
            src = row["src_path"]
            if inode_of(dst) is not None:
                continue
            if inode_of(src) is None:
                continue
            ensure_dir(os.path.dirname(dst))
            r = create_link(src, dst, fallback)
            if r.ok:
                fixed += 1
                db.execute(
                    "UPDATE links SET status='active', inode=?, missing_strikes=0 "
                    "WHERE id=?",
                    (inode_of(dst), row["id"]),
                )
            else:
                failed += 1
    if fixed:
        db.log_event("info", f"repair: re-created {fixed} link(s)")
    return {"fixed": fixed, "failed": failed}

File: backend-ts/src/api/links.ts
import { dirname } from 'node:path'
import { and, desc, eq, inArray, sql } from 'drizzle-orm'
import type { FastifyInstance } from 'fastify'
import type { Settings } from '../config/env.js'
import type { DbClient } from '../db/client.js'
import { logEvent } from '../db/events.js'
import { apps, links, rules as rulesTable } from '../db/schema.js'
import type { SettingsStore } from '../db/settings-store.js'
import {
  createLink,
  ensureDir,
  inodeOf,
  removeLink,
  resolveFsFallback,
} from '../core/fsutil.js'
import { normalizeFsFallback } from '../config/env.js'
import { HttpError } from '../http-error.js'
import { getCurrentUser } from './auth.js'

/** Browse/delete/repair hardlinks. Ported from api/links.py. Wire format is
 * snake_case, matching web/ and the Python backend. */

export interface LinksRouteOptions {
  db: DbClient
  settingsStore: SettingsStore
  env: Settings
}

type LinkRow = typeof links.$inferSelect

function linkOut(
  row: LinkRow,
  ruleName: string | null,
  appName: string | null,
  appType: string | null,
): Record<string, unknown> {
  return {
    id: row.id,
    rule_id: row.ruleId,
    app_id: row.appId,
    item_id: row.itemId,
    file_id: row.fileId,
    src_path: row.srcPath,
    dst_path: row.dstPath,
    inode: row.inode,
    status: row.status,
    created_at: row.createdAt,
    match_key: row.matchKey,
    missing_strikes: row.missingStrikes,
    rule_name: ruleName,
    app_name: appName,
    app_type: appType,
  }
}

export function registerLinksRoutes(app: FastifyInstance, opts: LinksRouteOptions): void {
  const { db, settingsStore, env } = opts

  app.get('/api/links', (request) => {
    getCurrentUser(request, db, settingsStore, env)
    const q = request.query as {
      app_id?: string
      rule_id?: string
      status?: string
      limit?: string
      offset?: string
    }
    const appId = q.app_id !== undefined ? Number(q.app_id) : undefined
    const ruleId = q.rule_id !== undefined ? Number(q.rule_id) : undefined
    const status = q.status !== undefined ? q.status : 'active'
    const limit = Math.min(
      5000,
      Math.max(1, q.limit !== undefined ? Number(q.limit) : 500),
    )
    const offset = Math.max(0, q.offset !== undefined ? Number(q.offset) : 0)

    const conditions = []
    if (appId !== undefined) conditions.push(eq(links.appId, appId))
    if (ruleId !== undefined) conditions.push(eq(links.ruleId, ruleId))
    if (status) conditions.push(eq(links.status, status))
    const where = conditions.length > 0 ? and(...conditions) : undefined

    const totalRow = db
      .select({ c: sql<number>`COUNT(*)` })
      .from(links)
      .where(where)
      .get()
    const total = totalRow?.c ?? 0

    const rows = db
      .select({
        link: links,
        ruleName: rulesTable.name,
        appName: apps.name,
        appType: apps.type,
      })
      .from(links)
      .leftJoin(rulesTable, eq(rulesTable.id, links.ruleId))
      .leftJoin(apps, eq(apps.id, links.appId))
      .where(where)
      .orderBy(desc(links.id))
      .limit(limit)
      .offset(offset)
      .all()

    return {
      items: rows.map((r) => linkOut(r.link, r.ruleName, r.appName, r.appType)),
      total,
      limit,
      offset,
    }
  })

  app.delete('/api/links/:linkId', (request, reply) => {
    getCurrentUser(request, db, settingsStore, env)
    const linkId = Number((request.params as { linkId: string }).linkId)
    const row = db.select().from(links).where(eq(links.id, linkId)).get()
    if (!row) throw new HttpError(404, 'link not found')
    const r = removeLink(row.dstPath)
    if (!r.ok && r.error) throw new HttpError(500, r.error)
    db.update(links).set({ status: 'missing' }).where(eq(links.id, linkId)).run()
    logEvent(db, 'info', `link removed: ${row.dstPath}`)
    reply.code(204).send()
  })

  // Re-create active links whose dst disappeared but whose source survives.
  app.post('/api/links/repair', (request) => {
    getCurrentUser(request, db, settingsStore, env)
    let fixed = 0
    let failed = 0
    const fallback = resolveFsFallback(settingsStore, normalizeFsFallback(env.fsFallback))
    // 'missing' too: source may have (re)appeared since removal.
    const rows = db
      .select({ id: links.id, dstPath: links.dstPath, srcPath: links.srcPath })
      .from(links)
      .where(inArray(links.status, ['active', 'stale', 'missing']))
      .all()
    for (const row of rows) {
      if (inodeOf(row.dstPath) !== null) continue
      if (inodeOf(row.srcPath) === null) continue
      ensureDir(dirname(row.dstPath))
      const r = createLink(row.srcPath, row.dstPath, fallback)
      if (r.ok) {
        fixed += 1
        db.update(links)
          .set({ status: 'active', inode: inodeOf(row.dstPath), missingStrikes: 0 })
          .where(eq(links.id, row.id))
          .run()
      } else {
        failed += 1
      }
    }
    if (fixed) logEvent(db, 'info', `repair: re-created ${fixed} link(s)`)
    return { fixed, failed }
  })
}
